### PR TITLE
Properly handle queue start/stop switch

### PR DESCRIPTION
**Pull request by thirtythreeforty**
This is a small series of patches to clean up the GUI callback for the queue worker.  It no longer deadlocks or crashes when cancelling the batch.  See #4882 for more details.

### CODE REVIEW

File: rtgui/batchqueue.cc [hunk 1]
```diff
@@ -778,7 +775,8 @@ rtengine::ProcessingJob* BatchQueue::imageReady(rtengine::IImagefloat* img)
     }
 
     redraw ();
-    notifyListener (queueEmptied);
+    const bool queueRunning = (processing != nullptr);
```
Comment (Floessie) [inline rtgui/batchqueue.cc]: Feel free to write `const bool queueRunning = processing`.

File: rtgui/batchqueuepanel.cc [hunk 2]
```diff
@@ -322,14 +326,9 @@ void BatchQueuePanel::addBatchQueueJobs(const std::vector<BatchQueueEntry*>& ent
...
-    if (qStartStop->get_active()) {
-        return true;
-    } else {
-        fdir->set_sensitive (true);
-        fformat->set_sensitive (true);
-        return false;
-    }
+    // This function is called from the background BatchQueue thread.
+    // It cannot call UI functions, so grab the stored state of qStartStop.
+    return qStartStopState;
```
Comment (Floessie) [inline rtgui/batchqueuepanel.cc]: :+1:

File: rtgui/batchqueuepanel.h [hunk 3]
```diff
@@ -53,29 +55,32 @@ class BatchQueuePanel : public Gtk::VBox,
 
     IdleRegister idle_register;
 
+    std::atomic<bool> qStartStopState;
```
Comment (Floessie) [inline rtgui/batchqueuepanel.h:56]: Excellent. Please move this before `idle_register`, which should always be the last member in case someone forgets to call `IdleRegister::destroy()` in the d'tor.

File: rtgui/batchqueuepanel.h [hunk 4]
```diff
@@ -60,22 +64,23 @@ class BatchQueuePanel : public Gtk::VBox,
     void init (RTWindow* parent);
 
     void addBatchQueueJobs(const std::vector<BatchQueueEntry*>& entries , bool head = false);
+    void saveOptions ();
+
+    bool handleShortcutKey (GdkEventKey* event);
```
Comment (Floessie) [inline rtgui/batchqueuepanel.h:69]: There must be a reason to make those accessible, but I can't see it from the diff.

Anyway, as one last nitpicking: Could you adapt the function order in the implementation to the one given in the header?

**Review by Floessie — APPROVED** [review]
The patch looks sane code-wise. I haven't tested it, but I really like the solution details and quick round trip.

**Comment by heckflosse** [comment]
I tested

1) no freezes :+1: 
2) when the queue is stopped while the second last image is processed, I can't start it again to process the last image.

File: rtgui/batchqueuepanel.h (hunk 4 above)
Comment (thirtythreeforty) [reply rtgui/batchqueuepanel.h:69]: There's only these two usages, I think:
https://github.com/Beep6581/RawTherapee/blob/b37d88fb2245fc99c529a377c57f3475c9da3c54/rtgui/rtwindow.cc#L674-L676
https://github.com/Beep6581/RawTherapee/blob/b37d88fb2245fc99c529a377c57f3475c9da3c54/rtgui/rtwindow.cc#L614-L616

**Comment by thirtythreeforty** [comment]
> when the queue is stopped while the second last image is processed, I can't start it again to process the last image.

Great catch, thank you.  The old code was (mis)using another GUI item's visibility as an indicator of idleness, which was now getting updated in the wrong order.

**Comment by heckflosse** [comment]
@thirtythreeforty Works fine now. Thank you. :+1: for merge from me

**Comment by heckflosse** [comment]
@thirtythreeforty Thank you very much for this fix :+1:

**Comment by thirtythreeforty** [comment]
Glad to contribute!

**Comment by ff2000** [comment]
I just pulled dev and rebuilt.
Unfortunately I can only start the queue if more than one image is there to being processed.
With just one image the start button is deactivated.
Auto start seems to work, it starts as soon as I add an image.

**Comment by heckflosse** [comment]
@ff2000 Confirmed. I missed that case in my tests :(

**Comment by thirtythreeforty** [comment]
There are a lot of edge cases with this :( I will look later!

**Comment by thirtythreeforty** [comment]
@ff2000 #4902 should fix this (bug source: my own carelessness). Would you please test?
